File: tests/lib/incident-loop.ts
import type { Page } from "@playwright/test";

export interface IncidentLoopOptions {
  /** 最长等待时间（毫秒），默认 8 分钟 */
  timeoutMs?: number;
  /** ask_human 时的自动回复文本 */
  askHumanReply: string;
}

export async function waitForIncidentResolution(
  page: Page,
  options: IncidentLoopOptions,
): Promise<void> {
  const { timeoutMs = 8 * 60 * 1000, askHumanReply } = options;
  const deadline = Date.now() + timeoutMs;
  const handledApprovals = new Set<string>();
  const handledQuestions = new Set<string>();

  while (Date.now() < deadline) {
    const summary = page.locator('[data-testid="summary-section"]');
    if (await summary.isVisible().catch(() => false)) break;

    const approvalCard = page.locator('[data-testid="approval-card"]').last();
    if (await approvalCard.isVisible().catch(() => false)) {
      const approvalId = await approvalCard.getAttribute("data-approval-id");
      const approvalKey = approvalId ?? "approval";
      if (!handledApprovals.has(approvalKey)) {
        const approveBtn = approvalCard.locator('[data-testid="approve-button"]');
        if (await approveBtn.isVisible().catch(() => false)) {
          handledApprovals.add(approvalKey);
          await approveBtn.click();
          await page.waitForTimeout(2000);
          continue;
        }
      }
    }

    const askHumanBanner = page.locator('[data-testid="ask-human-banner"]');
    if (await askHumanBanner.isVisible().catch(() => false)) {
      const question = (await askHumanBanner.textContent())?.trim() ?? "ask-human";
      if (!handledQuestions.has(question)) {
        handledQuestions.add(question);
        await page.locator('[data-testid="prompt-textarea"]').fill(askHumanReply);
        await page.locator('[data-testid="submit-incident"]').click();
        await page.waitForTimeout(2000);
        continue;
      }
    }

    await page.waitForTimeout(3000);
  }
}

File: test/e2e/lib/incident_loop.py
from __future__ import annotations

import time

from playwright.sync_api import Page


def wait_for_incident_resolution(
    page: Page,
    ask_human_reply: str,
    timeout_ms: int = 8 * 60 * 1000,
) -> None:
    deadline = time.monotonic() + timeout_ms / 1000
    handled_approvals: set[str] = set()
    handled_questions: set[str] = set()

    while time.monotonic() < deadline:
        summary = page.locator('[data-testid="summary-section"]')
        try:
            if summary.is_visible():
                break
        except Exception:
            pass

        approval_card = page.locator('[data-testid="approval-card"]').last
        try:
            if approval_card.is_visible():
                approval_id = approval_card.get_attribute("data-approval-id") or "approval"
                if approval_id not in handled_approvals:
                    approve_btn = approval_card.locator('[data-testid="approve-button"]')
                    try:
                        if approve_btn.is_visible():
                            handled_approvals.add(approval_id)
                            approve_btn.click()
                            page.wait_for_timeout(2000)
                            continue
                    except Exception:
                        pass
        except Exception:
            pass

        ask_human_banner = page.locator('[data-testid="ask-human-banner"]')
        try:
            if ask_human_banner.is_visible():
                question = (ask_human_banner.text_content() or "").strip() or "ask-human"
                if question not in handled_questions:
                    handled_questions.add(question)
                    page.locator('[data-testid="prompt-textarea"]').fill(ask_human_reply)
                    page.locator('[data-testid="submit-incident"]').click()
                    page.wait_for_timeout(2000)
                    continue
        except Exception:
            pass

        page.wait_for_timeout(3000)
